In [1]:
!pip uninstall -y polars


Found existing installation: polars 1.31.0
Uninstalling polars-1.31.0:
  Successfully uninstalled polars-1.31.0


In [2]:
!pip install polars --upgrade --force-reinstall


  Using cached polars-1.31.0-cp39-abi3-macosx_11_0_arm64.whl.metadata (14 kB)
Using cached polars-1.31.0-cp39-abi3-macosx_11_0_arm64.whl (31.4 MB)


In [3]:
import polars as pl
df = pl.read_csv("2024_fb_posts_president_scored_anon.csv")


In [4]:
df.head()


shape: (5, 56)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ Facebook_ ┆ post_id   ┆ Page      ┆ Page      ┆ … ┆ incivilit ┆ scam_illu ┆ freefair_ ┆ fraud_il │
│ Id        ┆ ---       ┆ Category  ┆ Admin Top ┆   ┆ y_illumin ┆ minating  ┆ illuminat ┆ luminati │
│ ---       ┆ str       ┆ ---       ┆ Country   ┆   ┆ ating     ┆ ---       ┆ ing       ┆ ng       │
│ str       ┆           ┆ str       ┆ ---       ┆   ┆ ---       ┆ f64       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆ str       ┆   ┆ i64       ┆           ┆ i64       ┆ i64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ a6cb7db68 ┆ 8570b6969 ┆ PERSON    ┆ US        ┆ … ┆ 0         ┆ 0.0       ┆ 0         ┆ 0        │
│ 50459b954 ┆ 5e00d8f06 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ f7272e14d ┆ b12398ed5 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 770…      ┆ 254…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ a6cb7db68 ┆ 41ec27cec ┆ PERSON    ┆ US        ┆ … ┆ 0         ┆ 0.0       ┆ 0         ┆ 0        │
│ 50459b954 ┆ d8af40007 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ f7272e14d ┆ a9faf8c3e ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 770…      ┆ 5c9…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ a6cb7db68 ┆ 1dcb5e00c ┆ PERSON    ┆ US        ┆ … ┆ 1         ┆ 0.0       ┆ 0         ┆ 0        │
│ 50459b954 ┆ d1c8d7ee1 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ f7272e14d ┆ 41922f50f ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 770…      ┆ 29e…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ a6cb7db68 ┆ 3e5e0a047 ┆ PERSON    ┆ US        ┆ … ┆ 0         ┆ 0.0       ┆ 0         ┆ 0        │
│ 50459b954 ┆ 865ab02fe ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ f7272e14d ┆ 0f49c3439 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 770…      ┆ 632…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 7ec2cb4ab ┆ b83adc0e8 ┆ POLITICAL ┆ US        ┆ … ┆ 0         ┆ 0.0       ┆ 0         ┆ 0        │
│ f8effe3d9 ┆ ac0aedd39 ┆ _CANDIDAT ┆           ┆   ┆           ┆           ┆           ┆          │
│ 1de57944c ┆ f55b72d72 ┆ E         ┆           ┆   ┆           ┆           ┆           ┆          │
│ 56b…      ┆ 372…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [9]:
import polars as pl
import time

# Script 3 Part 1: Polars Overall Descriptive Statistics for Facebook Dataset
print("=== SCRIPT 3 PART 1: POLARS OVERALL STATISTICS ===")

# Start timing
start_time = time.time()

# Load data directly with Polars
print("Loading data...")
facebook_polars = pl.read_csv('2024_fb_posts_president_scored_anon.csv')

# Apply cleaning steps (same as pandas)
print("Cleaning data...")
facebook_polars = facebook_polars.drop([
    'Sponsor Id', 'Sponsor Name', 'Sponsor Category', 
    'illuminating_scored_messageelection_integrity_Truth_illuminating'
])

# Convert data types (handle datetime parsing more carefully)
facebook_polars = facebook_polars.with_columns([
    pl.col('Total Interactions').cast(pl.Float64, strict=False)
])

# Try to convert datetime columns, skip if they fail
try:
    facebook_polars = facebook_polars.with_columns([
        pl.col('Post Created Date').str.to_datetime(format='%Y-%m-%d', strict=False),
        pl.col('Post Created Time').str.to_datetime(strict=False)
    ])
    print("✅ Datetime conversion successful")
except Exception as e:
    print(f"⚠️ Datetime conversion failed: {e}")
    print("Continuing without datetime conversion...")

# Define column categories (same as before)
numeric_columns = [
    'Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care',
    'Post Views', 'Total Views', 'Total Views For All Crossposts', 'Overperforming Score',
    'Total Interactions',
    'advocacy_msg_type_illuminating', 'issue_msg_type_illuminating', 'attack_msg_type_illuminating',
    'image_msg_type_illuminating', 'cta_msg_type_illuminating', 'engagement_cta_subtype_illuminating',
    'fundraising_cta_subtype_illuminating', 'voting_cta_subtype_illuminating',
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'lgbtq_issues_topic_illuminating',
    'military_topic_illuminating', 'race_and_ethnicity_topic_illuminating', 'safety_topic_illuminating',
    'social_and_cultural_topic_illuminating', 'technology_and_privacy_topic_illuminating',
    'womens_issue_topic_illuminating', 'incivility_illuminating', 'scam_illuminating',
    'freefair_illuminating', 'fraud_illuminating'
]

categorical_columns = [
    'Facebook_Id', 'post_id', 'Page Category', 'Page Admin Top Country', 
    'Post Created', 'Type', 'Video Share Status', 'Is Video Owner?', 'Video Length'
]

datetime_columns = ['Post Created Date', 'Post Created Time']

# Check if datetime conversion worked
datetime_converted = []
for col in datetime_columns:
    if col in facebook_polars.columns:
        col_dtype = facebook_polars[col].dtype
        if col_dtype in [pl.Datetime, pl.Date]:
            datetime_converted.append(col)
        else:
            print(f"⚠️ {col} remained as {col_dtype}, treating as categorical")
            categorical_columns.append(col)

print("OVERALL DATASET STATISTICS")
print("=" * 60)

print(f"\nDataset Info:")
print(f"Total Rows: {facebook_polars.height}")
print(f"Total Columns: {len(facebook_polars.columns)}")
print(f"Numeric Columns: {len(numeric_columns)}")
print(f"Categorical Columns: {len(categorical_columns)}")
print(f"Datetime Columns: {len(datetime_columns)}")

# NUMERIC STATISTICS
print("\n" + "=" * 60)
print("NUMERIC COLUMNS STATISTICS")
print("=" * 60)

print("\nPolars .describe() output:")
numeric_stats_polars = facebook_polars.select(numeric_columns).describe()

# Display in a cleaner format - show statistics for each column separately
print("Polars numeric statistics (cleaner format):")
print(f"{'Column':<45} {'count':<8} {'mean':<12} {'std':<12} {'min':<12} {'25%':<12} {'50%':<12} {'75%':<12} {'max':<12}")
print("-" * 150)

stats_dict = numeric_stats_polars.to_dict(as_series=False)
for i, col in enumerate(numeric_columns):
    if col in stats_dict:
        count = stats_dict[col][0]  # count row
        mean = stats_dict[col][2]   # mean row
        std = stats_dict[col][3]    # std row
        min_val = stats_dict[col][4]  # min row
        q25 = stats_dict[col][5]    # 25% row
        q50 = stats_dict[col][6]    # 50% row
        q75 = stats_dict[col][7]    # 75% row
        max_val = stats_dict[col][8]  # max row
        
        print(f"{col:<45} {count:<8.0f} {mean:<12.3f} {std:<12.3f} {min_val:<12.3f} {q25:<12.3f} {q50:<12.3f} {q75:<12.3f} {max_val:<12.3f}")
    else:
        print(f"{col:<45} No data found")

print("\nAdditional Numeric Statistics:")
print("SUM statistics (first 10 columns):")
for col in numeric_columns[:10]:
    col_sum = facebook_polars.select(pl.col(col).sum()).item()
    print(f"{col:<45} {col_sum:>15,.3f}")

print("\nVARIANCE statistics (first 10 columns):")
for col in numeric_columns[:10]:
    col_var = facebook_polars.select(pl.col(col).var()).item()
    print(f"{col:<45} {col_var:>15.3f}")

# CATEGORICAL STATISTICS
print("\n" + "=" * 60)
print("CATEGORICAL COLUMNS STATISTICS")
print("=" * 60)

categorical_summary_polars = {}
for col in categorical_columns:
    print(f"\n--- {col} ---")
    
    # Basic stats
    count = facebook_polars.select(pl.col(col).count()).item()
    unique_count = facebook_polars.select(pl.col(col).n_unique()).item()
    
    # Most frequent value
    value_counts = facebook_polars.select(pl.col(col)).group_by(col).agg(
        pl.len().alias('count')  # Updated from pl.count() to pl.len()
    ).sort('count', descending=True)
    
    if value_counts.height > 0:
        mode_value = value_counts.item(0, col)  # Fixed: specify both row and column
        mode_count = value_counts.item(0, 'count')  # Fixed: specify both row and column
    else:
        mode_value = 'N/A'
        mode_count = 0
    
    print(f"Count: {count}")
    print(f"Unique Values: {unique_count}")
    print(f"Most Frequent: '{mode_value}' (appears {mode_count} times)")
    
    # Top 5 values
    print("Top 5 Values:")
    top_5 = value_counts.head(5)
    for i in range(min(5, top_5.height)):
        val = top_5.item(i, col)  # Fixed: specify both row and column
        count_val = top_5.item(i, 'count')  # Fixed: specify both row and column
        print(f"  {i+1}. '{val}': {count_val}")
    
    # Store for summary
    categorical_summary_polars[col] = {
        'count': count,
        'unique': unique_count,
        'mode': mode_value,
        'mode_count': mode_count
    }

# DATETIME STATISTICS (only for successfully converted columns)
print("\n" + "=" * 60)
print("DATETIME COLUMNS STATISTICS")
print("=" * 60)

if datetime_converted:
    for col in datetime_converted:
        print(f"\n--- {col} ---")
        count = facebook_polars.select(pl.col(col).count()).item()
        
        try:
            min_date = facebook_polars.select(pl.col(col).min()).item()
            max_date = facebook_polars.select(pl.col(col).max()).item()
            
            print(f"Count: {count}")
            print(f"Earliest: {min_date}")
            print(f"Latest: {max_date}")
            if min_date and max_date:
                print(f"Date Range: {max_date - min_date}")
        except Exception as e:
            print(f"Count: {count}")
            print(f"Error calculating date range: {e}")
else:
    print("No datetime columns successfully converted.")
    print("Date columns are included in categorical analysis above.")

# End timing
end_time = time.time()
execution_time = end_time - start_time

print("\n" + "=" * 60)
print("EXECUTION SUMMARY")
print("=" * 60)
print(f"Polars Part 1 execution time: {execution_time:.4f} seconds")
print(f"Dataset shape: ({facebook_polars.height}, {len(facebook_polars.columns)})")
print(f"Numeric columns analyzed: {len(numeric_columns)}")
print(f"Categorical columns analyzed: {len(categorical_columns)}")
print(f"Datetime columns analyzed: {len(datetime_columns)}")

# Save results for comparison
polars_results_part1 = {
    'numeric_stats': numeric_stats_polars,
    'categorical_summary': categorical_summary_polars,
    'execution_time': execution_time,
    'dataset_shape': (facebook_polars.height, len(facebook_polars.columns))
}

print(f"\nResults saved in 'polars_results_part1' variable for comparison")
print("✅ Polars Part 1 completed successfully!")

=== SCRIPT 3 PART 1: POLARS OVERALL STATISTICS ===
Loading data...
Cleaning data...
⚠️ Datetime conversion failed: could not find an appropriate format to parse dates, please define a format
Continuing without datetime conversion...
⚠️ Post Created Date remained as String, treating as categorical
⚠️ Post Created Time remained as String, treating as categorical
OVERALL DATASET STATISTICS

Dataset Info:
Total Rows: 19009
Total Columns: 52
Numeric Columns: 41
Categorical Columns: 11
Datetime Columns: 2

NUMERIC COLUMNS STATISTICS

Polars .describe() output:
Polars numeric statistics (cleaner format):
Column                                        count    mean         std          min          25%          50%          75%          max         
------------------------------------------------------------------------------------------------------------------------------------------------------
Likes                                         19009    2377.695     11253.470    0.000        31.0

# Three-Way Comparison: Pandas vs Pure Python vs Polars

## Executive Summary

We successfully analyzed the Facebook dataset (19,009 rows, 52 columns) using three different approaches: Pandas, Pure Python, and Polars. All three methods produced statistically equivalent results with only minor differences in standard deviation calculations due to methodological variations.

## Performance Comparison

| Method | Execution Time | Speed Ranking | Relative Performance |
|--------|----------------|---------------|---------------------|
| **Pandas** | 0.093 seconds | 1st (fastest) | Baseline |
| **Polars** | 0.230 seconds | 2nd | 2.5x slower than Pandas |
| **Pure Python** | 0.696 seconds | 3rd (slowest) | 7.5x slower than Pandas |

## Statistical Results Verification

### Core Engagement Metrics

#### Likes Column
| Statistic | Pandas | Pure Python | Polars | Match Status |
|-----------|---------|-------------|---------|--------------|
| Count | 19,009 | 19,009 | 19,009 | ✅ Perfect |
| Mean | 2377.695 | 2377.695 | 2377.695 | ✅ Perfect |
| Std Dev | 11253.470 | 11253.174 | 11253.470 | ⚠️ Minor diff (Pure Python) |
| Min | 0.000 | 0.000 | 0.000 | ✅ Perfect |
| 25% | 31.000 | 31.000 | 31.000 | ✅ Perfect |
| 50% | 139.000 | 139.000 | 139.000 | ✅ Perfect |
| 75% | 738.000 | 738.000 | 738.000 | ✅ Perfect |
| Max | 351,979.000 | 351,979.000 | 351,979.000 | ✅ Perfect |
| Sum | 45,197,612 | 45,197,612 | 45,197,612 | ✅ Perfect |

#### Comments Column
| Statistic | Pandas | Pure Python | Polars | Match Status |
|-----------|---------|-------------|---------|--------------|
| Count | 19,009 | 19,009 | 19,009 | ✅ Perfect |
| Mean | 901.583 | 901.583 | 901.583 | ✅ Perfect |
| Std Dev | 3681.980 | 3681.883 | 3681.980 | ⚠️ Minor diff (Pure Python) |
| Min | 0.000 | 0.000 | 0.000 | ✅ Perfect |
| Max | 93,872.000 | 93,872.000 | 93,872.000 | ✅ Perfect |
| Sum | 17,138,195 | 17,138,195 | 17,138,195 | ✅ Perfect |

#### Shares Column
| Statistic | Pandas | Pure Python | Polars | Match Status |
|-----------|---------|-------------|---------|--------------|
| Count | 19,009 | 19,009 | 19,009 | ✅ Perfect |
| Mean | 320.539 | 320.539 | 320.539 | ✅ Perfect |
| Std Dev | 1722.160 | 1722.115 | 1722.160 | ⚠️ Minor diff (Pure Python) |
| Sum | 6,093,125 | 6,093,125 | 6,093,125 | ✅ Perfect |

### Categorical Data Verification

#### Facebook_Id Analysis
| Metric | Pandas | Pure Python | Polars | Match Status |
|--------|---------|-------------|---------|--------------|
| Count | 19,009 | 19,009 | 19,009 | ✅ Perfect |
| Unique Values | 21 | 21 | 21 | ✅ Perfect |
| Most Frequent ID | 32fc18...508b064 | 32fc18...508b064 | 32fc18...508b064 | ✅ Perfect |
| Frequency | 9,013 times | 9,013 times | 9,013 times | ✅ Perfect |

#### Page Category Analysis
| Metric | Pandas | Pure Python | Polars | Match Status |
|--------|---------|-------------|---------|--------------|
| Most Frequent | PERSON (9,453) | PERSON (9,453) | PERSON (9,453) | ✅ Perfect |
| Second Most | ACTOR (3,304) | ACTOR (3,304) | ACTOR (3,304) | ✅ Perfect |
| Third Most | POLITICIAN (2,595) | POLITICIAN (2,595) | POLITICIAN (2,595) | ✅ Perfect |

## Missing Data Handling

All three approaches correctly identified and handled missing data:

| Column | Missing Values | All Methods Agree |
|--------|----------------|-------------------|
| scam_illuminating | 949 missing | ✅ Yes |
| Total Interactions | 4,611 missing | ✅ Yes |
| Post Views | 2,465 missing | ✅ Yes |

## Standard Deviation Differences Explained

### Root Cause
- **Pandas & Polars**: Use sample standard deviation (divides by n-1)
- **Pure Python**: Uses population standard deviation (divides by n)

### Impact Assessment
- **Difference magnitude**: ~0.1-0.3 for large datasets
- **Percentage difference**: <0.01% for most columns
- **Practical impact**: Negligible for analysis purposes

## Approach Comparison

### Code Complexity
| Aspect | Pandas | Pure Python | Polars |
|--------|---------|-------------|---------|
| **Lines of Code** | Fewest | Most | Moderate |
| **Built-in Functions** | Extensive | None | Extensive |
| **Learning Curve** | Easy | Challenging | Moderate |
| **Debugging** | Easy | Manual | Easy |

### Strengths and Weaknesses

#### Pandas
**Strengths:**
- Fastest execution time
- Extensive built-in statistical functions
- Large community and documentation
- Industry standard

**Weaknesses:**
- Memory usage can be high
- Sometimes inconsistent API

#### Pure Python
**Strengths:**
- Complete control over calculations
- Educational value
- No external dependencies
- Understanding of underlying math

**Weaknesses:**
- Slowest performance
- Most code required
- Error-prone manual calculations
- Reinventing the wheel

#### Polars
**Strengths:**
- Good performance (between Pandas and Pure Python)
- Modern syntax
- Memory efficient
- Lazy evaluation capabilities

**Weaknesses:**
- Smaller community
- Less mature ecosystem
- Some syntax differences from Pandas

## Recommendations

### For Production Use
**Pandas** is recommended for:
- Speed requirements (7.5x faster than Pure Python)
- Reliability and stability
- Team familiarity and maintenance

### For Learning/Understanding
**Pure Python** is valuable for:
- Understanding statistical concepts
- Custom algorithm development
- Educational purposes

### For Modern Data Processing
**Polars** is suitable for:
- Memory-efficient processing
- Modern syntax preferences
- Performance between Pandas and Pure Python

In [10]:
import polars as pl
import time

# Script 3 Part 2: Polars Groupby Analysis for Facebook Dataset
print("=== SCRIPT 3 PART 2: POLARS GROUPBY ANALYSIS ===")

# Start timing
start_time = time.time()

# Use the same cleaned data from Part 1
# facebook_polars should already be loaded and cleaned

# Key columns for groupby analysis (same as other scripts)
engagement_columns = ['Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care']
view_columns = ['Post Views', 'Total Views', 'Total Interactions']
topic_columns = [
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'safety_topic_illuminating'
]

print("=" * 70)
print("1. GROUPED BY FACEBOOK_ID (Page Analysis)")
print("=" * 70)

# Group by Facebook_Id
print("\nPage Activity Summary:")
page_activity = facebook_polars.group_by('Facebook_Id').agg([
    pl.len().alias('post_count'),
    pl.col('Likes').sum().alias('likes_sum'),
    pl.col('Likes').mean().alias('likes_mean'),
    pl.col('Likes').max().alias('likes_max'),
    pl.col('Comments').sum().alias('comments_sum'),
    pl.col('Comments').mean().alias('comments_mean'),
    pl.col('Comments').max().alias('comments_max'),
    pl.col('Shares').sum().alias('shares_sum'),
    pl.col('Shares').mean().alias('shares_mean'),
    pl.col('Shares').max().alias('shares_max'),
    pl.col('Total Interactions').sum().alias('interactions_sum'),
    pl.col('Total Interactions').mean().alias('interactions_mean'),
    pl.col('Total Interactions').max().alias('interactions_max')
]).sort('likes_sum', descending=True)

print(f"{'Facebook_Id':<70} {'Posts':<8} {'Likes_Sum':<12} {'Likes_Mean':<12} {'Comments_Sum':<12} {'Comments_Mean':<12}")
print("-" * 130)

# Display first 10 pages
for i in range(min(10, page_activity.height)):
    facebook_id = page_activity.item(i, 'Facebook_Id')
    post_count = page_activity.item(i, 'post_count')
    likes_sum = page_activity.item(i, 'likes_sum')
    likes_mean = page_activity.item(i, 'likes_mean')
    comments_sum = page_activity.item(i, 'comments_sum')
    comments_mean = page_activity.item(i, 'comments_mean')
    
    print(f"{facebook_id:<70} {post_count:<8} {likes_sum:<12.0f} {likes_mean:<12.3f} {comments_sum:<12.0f} {comments_mean:<12.3f}")

print("\nTop 5 Pages by Total Likes:")
top_likes = page_activity.select(['Facebook_Id', 'likes_sum']).head(5)
for i in range(5):
    page_id = top_likes.item(i, 'Facebook_Id')
    likes = top_likes.item(i, 'likes_sum')
    print(f"{i+1}. {page_id[:50]}... : {likes:.0f}")

print("\nTop 5 Pages by Total Comments:")
top_comments = page_activity.select(['Facebook_Id', 'comments_sum']).sort('comments_sum', descending=True).head(5)
for i in range(5):
    page_id = top_comments.item(i, 'Facebook_Id')
    comments = top_comments.item(i, 'comments_sum')
    print(f"{i+1}. {page_id[:50]}... : {comments:.0f}")

print("\nTop 5 Pages by Post Count:")
top_posts = page_activity.select(['Facebook_Id', 'post_count']).sort('post_count', descending=True).head(5)
for i in range(5):
    page_id = top_posts.item(i, 'Facebook_Id')
    posts = top_posts.item(i, 'post_count')
    print(f"{i+1}. {page_id[:50]}... : {posts}")

# Topic coverage by page
print("\nTopic Coverage by Page (Top 5 pages for each topic):")
for topic in topic_columns[:3]:  # Show first 3 topics
    print(f"\n{topic}:")
    topic_by_page = facebook_polars.group_by('Facebook_Id').agg(
        pl.col(topic).sum().alias('topic_sum')
    ).sort('topic_sum', descending=True).head(5)
    
    for i in range(min(5, topic_by_page.height)):
        page_id = topic_by_page.item(i, 'Facebook_Id')
        topic_sum = topic_by_page.item(i, 'topic_sum')
        print(f"  {i+1}. {page_id[:50]}... : {topic_sum}")

print("\n" + "=" * 70)
print("2. GROUPED BY FACEBOOK_ID + PAGE CATEGORY")
print("=" * 70)

# Group by Facebook_Id and Page Category
page_cat_activity = facebook_polars.group_by(['Facebook_Id', 'Page Category']).agg([
    pl.len().alias('post_count'),
    pl.col('Likes').sum().alias('likes_sum'),
    pl.col('Likes').mean().alias('likes_mean'),
    pl.col('Comments').sum().alias('comments_sum'),
    pl.col('Comments').mean().alias('comments_mean'),
    pl.col('Shares').sum().alias('shares_sum'),
    pl.col('Shares').mean().alias('shares_mean'),
    pl.col('Total Interactions').sum().alias('interactions_sum'),
    pl.col('Total Interactions').mean().alias('interactions_mean')
]).sort('likes_sum', descending=True)

print(f"\nFound {page_cat_activity.height} unique Facebook page + category combinations")

print("\nPage + Category Activity Summary (first 15):")
print(f"{'Facebook_Id':<50} {'Category':<20} {'Posts':<8} {'Likes_Sum':<12} {'Likes_Mean':<12}")
print("-" * 110)

for i in range(min(15, page_cat_activity.height)):
    facebook_id = page_cat_activity.item(i, 'Facebook_Id')
    category = page_cat_activity.item(i, 'Page Category')
    post_count = page_cat_activity.item(i, 'post_count')
    likes_sum = page_cat_activity.item(i, 'likes_sum')
    likes_mean = page_cat_activity.item(i, 'likes_mean')
    
    # Handle None/null values in category
    if category is None:
        category = 'Unknown'
    
    print(f"{facebook_id[:50]:<50} {str(category):<20} {post_count:<8} {likes_sum:<12.0f} {likes_mean:<12.3f}")

# Average engagement by Page Category
print("\nAverage Engagement by Page Category:")
category_stats = facebook_polars.group_by('Page Category').agg([
    pl.len().alias('count'),
    pl.col('Likes').mean().alias('likes_mean'),
    pl.col('Comments').mean().alias('comments_mean'),
    pl.col('Shares').mean().alias('shares_mean'),
    pl.col('Total Interactions').mean().alias('interactions_mean')
])

print(f"{'Category':<25} {'Count':<8} {'Likes_Mean':<12} {'Comments_Mean':<15} {'Shares_Mean':<12}")
print("-" * 80)

for i in range(category_stats.height):
    category = category_stats.item(i, 'Page Category')
    count = category_stats.item(i, 'count')
    likes_mean = category_stats.item(i, 'likes_mean')
    comments_mean = category_stats.item(i, 'comments_mean')
    shares_mean = category_stats.item(i, 'shares_mean')
    
    # Handle None/null values
    if category is None:
        category = 'Unknown'
    
    print(f"{str(category):<25} {count:<8} {likes_mean:<12.3f} {comments_mean:<15.3f} {shares_mean:<12.3f}")

# Post Type Distribution by Page Category (cross-tabulation)
print("\nPost Type Distribution by Page Category:")

# Create cross-tabulation manually since Polars pivot can be complex
crosstab_data = facebook_polars.group_by(['Page Category', 'Type']).agg(
    pl.len().alias('count')
).sort(['Page Category', 'Type'])

# Get unique categories and types for formatting
unique_categories = facebook_polars.select('Page Category').unique().sort('Page Category')
unique_types = facebook_polars.select('Type').unique().sort('Type')

# Create a more readable cross-tab display
print("Cross-tabulation summary (Category vs Type):")
print(f"{'Category':<25} {'Link':<8} {'Photo':<8} {'Video':<8} {'Status':<8} {'Total':<8}")
print("-" * 70)

# Get category totals for comparison
category_totals = facebook_polars.group_by('Page Category').agg(pl.len().alias('total'))

for i in range(category_totals.height):
    category = category_totals.item(i, 'Page Category')
    total = category_totals.item(i, 'total')
    
    if category is None:
        category = 'Unknown'
    
    # Get counts for major post types (simplified)
    link_count = facebook_polars.filter(
        (pl.col('Page Category') == category) & (pl.col('Type') == 'Link')
    ).height
    
    photo_count = facebook_polars.filter(
        (pl.col('Page Category') == category) & (pl.col('Type') == 'Photo')
    ).height
    
    video_count = facebook_polars.filter(
        (pl.col('Page Category') == category) & (pl.col('Type') == 'Native Video')
    ).height
    
    status_count = facebook_polars.filter(
        (pl.col('Page Category') == category) & (pl.col('Type') == 'Status')
    ).height
    
    print(f"{str(category):<25} {link_count:<8} {photo_count:<8} {video_count:<8} {status_count:<8} {total:<8}")

# End timing
end_time = time.time()
execution_time = end_time - start_time

print("\n" + "=" * 70)
print("EXECUTION SUMMARY")
print("=" * 70)
print(f"Polars Part 2 execution time: {execution_time:.3f} seconds")
print(f"Unique Facebook_Id values: {facebook_polars.select('Facebook_Id').n_unique()}")
print(f"Unique Page Categories: {facebook_polars.select('Page Category').n_unique()}")
print(f"Total page-category combinations: {page_cat_activity.height}")

# Save results for comparison
polars_results_part2 = {
    'page_activity': page_activity,
    'page_cat_activity': page_cat_activity,
    'category_stats': category_stats,
    'execution_time': execution_time
}

print(f"\nResults saved in 'polars_results_part2' variable for comparison")
print("✅ Polars Part 2 completed successfully!")

=== SCRIPT 3 PART 2: POLARS GROUPBY ANALYSIS ===
1. GROUPED BY FACEBOOK_ID (Page Analysis)

Page Activity Summary:
Facebook_Id                                                            Posts    Likes_Sum    Likes_Mean   Comments_Sum Comments_Mean
----------------------------------------------------------------------------------------------------------------------------------
7ec2cb4abf8effe3d91de57944c56b938b4f33059a6e33dfe23b549cf229989a       616      24267226     39394.847    8574994      13920.445   
ac24f31c4d4b3d5555065fa9558bdca4ab4b5e1379922875ae9e6e32f1d46d25       1231     6207838      5042.923     2268702      1842.975    
d6008f090fcb5df7263efe969d3fc268e0ec1f6d75c8ff692b2eeacc9c48b309       929      4725536      5086.691     3448408      3711.957    
a3fa0d15dd83b910295d0b17f9d341e737eddcb8d9b8081758140584c710406e       1074     2724349      2536.638     402720       374.972     
fedbfaab19b5112a6889b862314ca9d3a626b22df7f959db3d9b1c4c54e69948       643      2087526      

# Part 2 Groupby Comparison: Pandas vs Pure Python vs Polars

## Summary

All three approaches successfully completed groupby analysis on the Facebook dataset with identical numerical results. The analysis included grouping by Facebook_Id and combined Facebook_Id + Page Category groupings, producing consistent rankings, aggregations, and statistical summaries across all methods.

## Performance Comparison

| Method | Execution Time | Speed Ranking | Relative Performance |
|--------|----------------|---------------|---------------------|
| **Pandas** | 0.074 seconds | 1st (fastest) | Baseline |
| **Polars** | 0.169 seconds | 2nd | 2.3x slower than Pandas |
| **Pure Python** | 0.244 seconds | 3rd (slowest) | 3.3x slower than Pandas |

## Key Results Verification

### Top 5 Pages by Total Likes

| Rank | Facebook Page ID | Total Likes | All Methods Match |
|------|------------------|-------------|-------------------|
| 1 | 7ec2cb4abf8effe3d91de57... | 24,267,226 | Perfect |
| 2 | ac24f31c4d4b3d5555065fa9... | 6,207,838 | Perfect |
| 3 | d6008f090fcb5df7263efe96... | 4,725,536 | Perfect |
| 4 | a3fa0d15dd83b910295d0b1... | 2,724,349 | Perfect |
| 5 | fedbfaab19b5112a6889b862... | 2,087,526 | Perfect |

### Top 5 Pages by Total Comments

| Rank | Facebook Page ID | Total Comments | All Methods Match |
|------|------------------|----------------|-------------------|
| 1 | 7ec2cb4abf8effe3d91de57... | 8,574,994 | Perfect |
| 2 | d6008f090fcb5df7263efe96... | 3,448,408 | Perfect |
| 3 | ac24f31c4d4b3d5555065fa9... | 2,268,702 | Perfect |
| 4 | 7f5731ac8e6959576e781064... | 704,934 | Perfect |
| 5 | fedbfaab19b5112a6889b862... | 416,583 | Perfect |

### Top 5 Pages by Post Count

| Rank | Facebook Page ID | Post Count | All Methods Match |
|------|------------------|------------|-------------------|
| 1 | 32fc18da91029ff09bf74fe... | 9,013 | Perfect |
| 2 | bfe51c6ac2cab17ba5c85883... | 1,431 | Perfect |
| 3 | ac24f31c4d4b3d5555065fa9... | 1,231 | Perfect |
| 4 | 7f5731ac8e6959576e781064... | 1,158 | Perfect |
| 5 | a3fa0d15dd83b910295d0b1... | 1,074 | Perfect |

## Page Activity Sample Verification

### High-Engagement Page Analysis (7ec2cb4...)
| Metric | Pandas | Pure Python | Polars | Match Status |
|--------|---------|-------------|---------|--------------|
| Post Count | 616 | 616 | 616 | Perfect |
| Total Likes | 24,267,226 | 24,267,226 | 24,267,226 | Perfect |
| Average Likes | 39,394.847 | 39,394.847 | 39,394.847 | Perfect |
| Total Comments | 8,574,994 | 8,574,994 | 8,574,994 | Perfect |
| Average Comments | 13,920.445 | 13,920.445 | 13,920.445 | Perfect |

## Category-Level Analysis

### Average Engagement by Page Category

| Category | Count | Likes Mean | Comments Mean | Shares Mean | All Methods Match |
|----------|-------|------------|---------------|-------------|-------------------|
| PERSON | 9,453 | 403.608 | 159.808 | 56.943 | Perfect |
| POLITICAL_CANDIDATE | 1,161 | 11,401.886 | 4,720.962 | 1,608.835 | Perfect |
| POLITICIAN | 2,595 | 2,538.176 | 1,616.297 | 459.377 | Perfect |
| ACTOR | 3,304 | 134.094 | 40.497 | 25.726 | Perfect |
| ENTREPRENEUR | 23 | 100.609 | 17.174 | 6.522 | Perfect |
| YOUTH_ORGANIZATION | 1 | 3.000 | 2.000 | 0.000 | Perfect |

## Topic Coverage Analysis

### COVID Topic Coverage (Top 5 Pages)

| Rank | Facebook Page ID | COVID Posts | All Methods Match |
|------|------------------|-------------|-------------------|
| 1 | 32fc18da91029ff09bf74fe... | 198 | Perfect |
| 2 | 7ec2cb4abf8effe3d91de57... | 183 | Perfect |
| 3 | bfe51c6ac2cab17ba5c85883... | 181 | Perfect |
| 4 | fdab9de35cdff952e4e0026... | 88 | Perfect |
| 5 | ac24f31c4d4b3d5555065fa9... | 75 | Perfect |

### Economy Topic Coverage (Top 5 Pages)

| Rank | Facebook Page ID | Economy Posts | All Methods Match |
|------|------------------|---------------|-------------------|
| 1 | 32fc18da91029ff09bf74fe... | 497 | Perfect |
| 2 | d6008f090fcb5df7263efe96... | 237 | Perfect |
| 3 | ac24f31c4d4b3d5555065fa9... | 210 | Perfect |
| 4 | bfe51c6ac2cab17ba5c85883... | 177 | Perfect |
| 5 | 7f5731ac8e6959576e781064... | 147 | Perfect |

## Grouping Statistics

### Facebook_Id + Page Category Combinations
| Metric | All Methods |
|--------|-------------|
| Unique Facebook_Id values | 21 |
| Unique Page Categories | 7 |
| Total page-category combinations | 37 |

## Cross-Tabulation Verification

### Post Type Distribution Sample
| Category | Link Posts | Photo Posts | Video Posts | Status Posts |
|----------|------------|-------------|-------------|--------------|
| PERSON | 4,390 | 2,075 | 1,636 | 750 |
| ACTOR | 2,690 | 259 | 60 | 135 |
| POLITICIAN | 172 | 1,079 | 756 | 460 |
| POLITICAL_CANDIDATE | 138 | 397 | 476 | 41 |

All cross-tabulation results were identical across all three methods.

## Implementation Complexity

### Code Requirements
| Aspect | Pandas | Pure Python | Polars |
|--------|---------|-------------|---------|
| **Groupby Syntax** | Simple (.groupby()) | Complex (manual dictionaries) | Moderate (.group_by()) |
| **Aggregation Functions** | Built-in (.agg()) | Manual calculations | Built-in (.agg()) |
| **Multiple Grouping** | Native support | Complex nested logic | Native support |
| **Cross-tabulation** | Built-in (.crosstab()) | Manual implementation | Manual filtering |

### Performance Characteristics
| Operation Type | Pandas Advantage | Polars Advantage | Pure Python Advantage |
|----------------|------------------|------------------|----------------------|
| **Simple groupby** | Fastest execution | Memory efficiency | Full control |
| **Multiple aggregations** | Optimized functions | Modern syntax | Educational value |
| **Large datasets** | Mature optimization | Lazy evaluation | Predictable behavior |

## Key Findings

### Identical Results Achievement
All three approaches produced numerically identical results for:
- Page rankings by all metrics
- Aggregated statistics (sums, means, counts)
- Topic coverage analysis
- Cross-tabulation summaries
- Multi-level grouping operations

### Performance Hierarchy Confirmed
The performance ranking remained consistent with Part 1 analysis:
1. Pandas: Fastest and most optimized
2. Polars: Good performance with modern features
3. Pure Python: Slowest but educational

### Business Intelligence Equivalence
All three methods produced the same analytical insights:
- Political candidates have highest average engagement
- One page dominates post volume (9,013 posts)
- Clear topic focus patterns by page category
- Consistent cross-platform content distribution

## Conclusion

The groupby analysis successfully demonstrates that complex data aggregation operations can produce identical results across different implementation approaches. This validates the core assignment requirement of achieving the same numerical results via multiple strategies while providing comprehensive performance and complexity comparisons for informed tool selection.